src/target/common/devo/protocol/pwm.c

```diff
@@ -79,29 +68,117 @@ void PWM_Stop()
         return;
 
     timer_disable_counter(TIM1);
-    nvic_disable_irq(NVIC_TIM1_UP_IRQ);
-    timer_disable_oc_output(TIM1, _PWM_TIM_OC);
-    rcc_peripheral_disable_clock(&RCC_APB2ENR, RCC_APB2ENR_TIM1EN);
+    nvic_disable_irq(_PWM_NVIC_DMA_CHANNEL_IRQ);
+    dma_disable_transfer_complete_interrupt(_PWM_DMA, _PWM_DMA_CHANNEL);
+    dma_disable_channel(_PWM_DMA, _PWM_DMA_CHANNEL);
+    rcc_peripheral_disable_clock(&RCC_APB2ENR, RCC_APB2ENR_TIM1EN | _RCC_AHBENR_DMAEN);
 
 #if _PWM_PIN == GPIO_USART1_TX
     UART_Initialize();
 #endif
 }
 
-extern volatile u16 *pwm;  // defined in pwm_rom.c
-void PPM_Enable(unsigned low_time, volatile u16 *pulses)
+void PPM_Enable(unsigned low_time, volatile u16 *pulses, u8 num_pulses)
+{
+    if (!pulses || num_pulses < 1) return;
+
+    // Setup DMA
+    dma_channel_reset(_PWM_DMA, _PWM_DMA_CHANNEL);
+    dma_set_peripheral_address(_PWM_DMA, _PWM_DMA_CHANNEL, (u32) &_PWM_TIM_ARR);
+    dma_set_memory_address(_PWM_DMA, _PWM_DMA_CHANNEL, (u32) pulses);
+    dma_set_number_of_data(_PWM_DMA, _PWM_DMA_CHANNEL, num_pulses);
+    dma_set_read_from_memory(_PWM_DMA, _PWM_DMA_CHANNEL);
+    dma_enable_memory_increment_mode(_PWM_DMA, _PWM_DMA_CHANNEL);
+    dma_set_peripheral_size(_PWM_DMA, _PWM_DMA_CHANNEL, DMA_CCR_PSIZE_16BIT);
+    dma_set_memory_size(_PWM_DMA, _PWM_DMA_CHANNEL, DMA_CCR_MSIZE_16BIT);
+    dma_set_priority(_PWM_DMA, _PWM_DMA_CHANNEL, DMA_CCR_PL_VERY_HIGH);
+    dma_enable_channel(_PWM_DMA, _PWM_DMA_CHANNEL);    // dma ready to go
+
+    // Setup timer for PPM
+    timer_set_oc_value(TIM1, _PWM_TIM_OC, low_time);
+    timer_set_period(TIM1, 22500);
+    timer_set_oc_polarity_low(TIM1, _PWM_TIM_OC);       // output active low
+    timer_set_oc_mode(TIM1, _PWM_TIM_OC, TIM_OCM_PWM1); // output active when timer below compare
+    timer_enable_oc_output(TIM1, _PWM_TIM_OC);          // enable OCx to pin
+    timer_enable_break_main_output(TIM1);               // master output enable
+    timer_set_dma_on_update_event(TIM1);
+    timer_enable_irq(TIM1, _PWM_TIM_DIER_DMAEN); // enable timer dma request (despite function name)
+    timer_generate_event(TIM1, TIM_EGR_UG);      // Generate update event to start DMA
+    timer_enable_counter(TIM1);
+}
+
+#ifndef MODULAR
+#define PXX_PKT_BYTES 18
+#define PXX_PKT_BITS  (20 * 8 + 32)  // every 5th bit might be escaped
+static const u8 pxx_flag[] = {15, 23, 23, 23, 23, 23, 23, 15};  // 7e
+static u8 pxx_bits[PXX_PKT_BITS];
+void PXX_Enable(u8 *packet)
```

Comment:
why this function is defined here instead of in the same file as caller?

Reply:
Follows the example of PPM_Enable.  Functions with dependencies on the libopencm3 library should be in the target/ tree.

Considered putting the bytes-to-bits conversion in the protocol file but it fits better here with the dma initialization.

Reply:
i see .